plugins/chat/app/services/chat/lookup_channel_threads.rb

```diff
@@ -54,28 +54,87 @@ def can_view_channel(guardian:, channel:, **)
     end
 
     def fetch_threads(guardian:, channel:, **)
-      Chat::Thread
-        .strict_loading
-        .includes(
-          :channel,
-          last_reply: %i[user uploads],
-          original_message_user: :user_status,
-          original_message: [
-            :chat_webhook_event,
-            :chat_mentions,
-            :chat_channel,
-            user: :user_status,
-          ],
+      threads =
+        Chat::Thread
+          .strict_loading
+          .includes(
+            :channel,
+            :user_chat_thread_memberships,
+            original_message_user: :user_status,
+            original_message: [
+              :chat_webhook_event,
+              :chat_mentions,
+              :chat_channel,
+              user: :user_status,
+            ],
+          )
+          .joins(
+            "JOIN (#{tracked_threads_subquery(guardian, channel)}) tracked_threads_subquery
+              ON tracked_threads_subquery.thread_id = chat_threads.id",
+          )
+          .joins(:user_chat_thread_memberships)
+          .where(user_chat_thread_memberships_chat_threads: { user_id: guardian.user.id })
+          .order(<<~SQL)
+            CASE WHEN user_chat_thread_memberships_chat_threads.last_read_message_id IS NULL
+              OR tracked_threads_subquery.latest_message_id >
+                user_chat_thread_memberships_chat_threads.last_read_message_id THEN 0 ELSE 1 END,
+              tracked_threads_subquery.latest_message_created_at DESC
```

Comment:
I do have some concerns about the ordering here since there isn't an index that can cover this so the query can become  less efficient if a user is tracking alot of threads. Is there some sort of upper limit for the number of threads a user can track?

Reply:
There isn't at the moment, however we do limit this query to 50 results. Maybe we could also limit the `tracked_threads_subquery` to only get the last N created threads? I also think we could probably just leave it for now and come back to it when the performance becomes more of an issue.

Though I am interested in what index we could add here?

Reply:
I don't think the limit helps here because PG needs to first fetch all the associated records, order them before picking out the first 50 records. 

> Though I am interested in what index we could add here?

With the conditional in the ordering clause, I don't think there is an index we can add here. The only way I think we can ensure this query continues to be performant even when the user has a large number of threads is to break the query into multiple queries. 

1. First query fetches `user_chat_thread_memberships_chat_threads.last_read_message_id IS NULL` 
2. Second query fetches `tracked_threads_subquery.latest_message_id > user_chat_thread_memberships_chat_threads.last_read_message_id`
3. Final query fetches the rest of the messages. 

Each query will limit it to 50 records and once we have 50 records, we don't need to execute the subsequent queries. Worst case is that we end up with 3 queries but I don't think it is that bad if the queries are fairly efficient.

Reply:
Oh right I see what you are saying. However I think we can just split it into 2 and say `user_chat_thread_memberships_chat_threads.last_read_message_id IS NULL OR tracked_threads_subquery.latest_message_id > user_chat_thread_memberships_chat_threads.last_read_message_id` since they are both just saying "get the unread threads", and then the second one is the read threads. Will just do that and push it up for you to see

Reply:
@martin-brennan O nice :+1: The two new query looks much better and easier to understand as well :+1: